### ISSUE TITLE
Upgrade from Drupal 7: Take Backdrop files directory into account

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#3706]
**Description of the bug**
After an upgrade of a Drupal 7 site to Backdrop, I saw several "failed to open stream" messages regarding public files; public images of the site were not displayed. I checked the page `admin/config/media/file-system` and saw that the public file system path was `sites/default/files`. When I changed it to `files`, the messages went away, and public images were displayed.

**Steps To Reproduce**
1. Prepare an upgrade of a D7 site to Backdrop :-)
2. As one step, copy the files directory from Drupal 7 (usually `drupal/sites/default/files`) into the Backdrop site (usually `backdrop/files`), as advised in the [Upgrade instructions](https://backdropcms.org/upgrade-from-drupal).
3. Finish the upgrade preparation.
4. Run `update.php` on the Backdrop site.
5. Go to the home page of your Backdrop site, or to any page which uses public files, e.g. images.

**Actual behavior**
- Public images aren't visible on the home page.
  (Under certain circumstances, you may see several "failed to open stream" messages.)
- On `admin/config/media/file-system` the public file system path is the same as in your D7 site, usually `sites/default/files`.

**Expected behavior**
- Public images are visible.
- The public file system path on `admin/config/media/file-system` should be `files`.
- If the *private* file system path was also within the Drupal directory, e.g. `sites/default/files/private`, it should be changed for Backdrop respectively.

**Comment by herbdool** [comment]
This is annoying but I'm not sure what can done. The setting exists in a config file and settings need to be changed manually in the UI. I'm not sure if "autodiscover" would be feasible.

**Comment by jenlampton** [comment]
Yes, this is a documentation issue. Closing this issue (and I'll fix the documentation).

**Comment by laryn** [comment]
One possibility would be to display a message with a link to the media/file settings page and instructions on what they need to do, especially if the file directory is set to `sites/default/files` as it will usually be in process of upgrade. (ie. bring the documentation directly to the user that needs it)

**Comment by jenlampton** [comment]
I'm not sure there's any way for us to tell if the user wanted the files directory to be `sites/default/files` or not. There are plenty of instances where files are requested but missing, that might trigger a false positive.

**Comment by laryn** [comment]
Is there a way to tell whether this is the final steps of an upgrade from Drupal 7 as opposed to just any old time where a file is requested and not available? If we know we're just completing an upgrade and we also know that the filesystem is set to `sites/default/files` I think a friendly note about this would be very helpful to people. Just spitballing.

**Comment by herbdool** [comment]
Perhaps it could check a record in file_managed during an upgrade and if it can't be found to then look in the standard location in Backdrop, then update the file path setting if found there. It should only try once though in case they want to put it elsewhere.

**Comment by jenlampton** [comment]
The file_managed table doesn't contain the file path, but it sometimes
contains info about files that no longer exist on disk (especially the
oldest ones).

I don't think this is something we can make a safe assumption about, but I
do think adding a warning is a good idea.

**Comment by olafgrabienski** [comment]
Reopening this issue because we're discussing ideas how to fix or mitigate it 😄

**Comment by olafgrabienski** [comment]
Hm, didn't hit `Reopen and comment`, let's try again!

**Comment by olafgrabienski** [comment]
> The file_managed table doesn't contain the file path, but it sometimes
contains info about files that no longer exist on disk ...

@jenlampton Is there another place where the information about actual file paths is stored? If it doesn't exist or isn't yet easily visible, I'd suggest to collect the information and make it centrally available, e.g. on `admin/content/files`. Based on this information, we could offer people coming from Drupal to bulk update some or all of their file paths.

**Comment by jenlampton** [comment]
> Is there another place where the information about actual file paths is stored?

The "File directory path" is stored in config (it was previously a variable in Drupal). This is usually `sites/default/files` for Drupal and `files` for Backdrop.

A path relative to the "File directory path" is stored in the file_managed table. This could be something like `images/logos/sample.png`.

The complete path to any individual file is not stored anywhere. It's assembled on-the-fly by adding the "File directory path" to the relative location of the file, so:

URL to file = `sites/default/files` + `/` + `images/logos/sample.png`

This makes our job of guessing the file directory path hard. 

My preference would be to convert the variable to config, like we are doing now, and then throw a warning on the status report if the file directory path contains `sites` and an error if it contains `sites/default`.

**Comment by klonos** [comment]
> My preference would be to convert the variable to config, like we are doing now, and then throw a warning on the status report if the file directory path contains `sites` and an error if it contains `sites/default`.

Still trying to fully get my head around this, but all that sounds reasonable to me.
